File: app/src/main/java/com/hustunique/morii/home/MusicDiaryAdapter.java
package com.hustunique.morii.home;

import static com.hustunique.morii.util.MyApplication.context;
import static com.hustunique.morii.util.MyApplication.musicTabList;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.ActivityOptions;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.Drawable;
import android.util.Log;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.imageview.ShapeableImageView;
import com.hustunique.morii.content.ContentActivity;
import com.hustunique.morii.util.DatabaseUtil;

import java.io.File;
import java.util.List;
import java.util.Locale;

import morii.R;


/*
① 创建一个继承RecyclerView.Adapter<VH>的Adapter类
② 创建一个继承RecyclerView.ViewHolder的静态内部类
③ 在Adapter中实现3个方法：
   onCreateViewHolder()
   onBindViewHolder()
   getItemCount()
*/
public class MusicDiaryAdapter extends RecyclerView.Adapter<MusicDiaryAdapter.MyViewHolder> {

    private final Activity activity;
    private static final String TAG = "MusicDiaryAdapter";

    private final List<MusicDiaryItem> list;
    //类型待定


    //构造方法，传入数据
    public MusicDiaryAdapter(Activity context, List<MusicDiaryItem> list) {
        this.activity = context;
        this.list = list;
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        //创建ViewHolder，返回每一项的布局
        View inflater = LayoutInflater.from(activity).inflate(R.layout.cardview_item, parent, false);
        return new MyViewHolder(inflater);
    }

    @Override
    public void onBindViewHolder(MyViewHolder holder, @SuppressLint("RecyclerView") int position) {
        //将数据和控件绑定
        //
        MusicDiaryItem musicDiaryItem = list.get(position);
        String imagePath = musicDiaryItem.getImagePath();
        if (null != imagePath) {
            Glide.with(holder.itemView).load(imagePath)
                    .into(holder.PhotoTitle);
            Log.d(TAG, "load image success");
        } else {
            Glide.with(holder.itemView)
                    .load(musicTabList.get(musicDiaryItem.getMusicTabId()).getImageResId())
                    .into(holder.PhotoTitle);
            Log.d(TAG, "onBindViewHolder: " + musicTabList.get(musicDiaryItem.getMusicTabId()).getImageResId());
        }
        holder.TextTitle.setText(musicDiaryItem.getTitle());
        holder.TextDate.setText("# " + musicDiaryItem.getDate());
        Log.d("RECYCLER", String.valueOf(position));
        holder.itemView.setOnClickListener(v -> {
            Intent intent = new Intent(activity, ContentActivity.class);
            intent.putExtra("diary", musicDiaryItem);
            ActivityOptions options = ActivityOptions
                    .makeSceneTransitionAnimation((Activity) activity,
                            Pair.create(holder.PhotoTitle, "photo"));
            activity.startActivity(intent, options.toBundle());
        });
        holder.itemView.setOnLongClickListener(v -> {
            MaterialAlertDialogBuilder builder = new MaterialAlertDialogBuilder(activity);
            builder.setTitle("确定删除这个音乐日记吗？")
                    .setMessage("这个操作不可被撤销。")
                    .setNegativeButton("取消", (dialog, which) -> {
                    })
                    .setPositiveButton("确认", (dialog, which) -> {
                        DatabaseUtil.deleteDiary(musicDiaryItem.getItemID());
                        Log.d(TAG, "deletePosition: " + holder.getLayoutPosition());
                        notifyItemRemoved(holder.getLayoutPosition());
                        list.remove(holder.getLayoutPosition());
                    })
                    .show();
            return true;
        });
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    //内部类，绑定控件
    static class MyViewHolder extends RecyclerView.ViewHolder {
        TextView TextTitle;
        TextView TextDate;
        ShapeableImageView PhotoTitle;

        public MyViewHolder(View itemView) {
            super(itemView);
            TextTitle = itemView.findViewById(R.id.TextTitle);
            TextDate = itemView.findViewById(R.id.TextDate);
            PhotoTitle = itemView.findViewById(R.id.PhotoTitle);

        }
    }
}

File: app/src/main/java/com/hustunique/morii/home/MusicDiaryAdapter.kt
package com.hustunique.morii.home

import android.annotation.SuppressLint
import android.app.Activity
import android.app.ActivityOptions
import android.content.DialogInterface
import android.content.Intent
import android.util.Log
import android.util.Pair
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.hustunique.morii.content.ContentActivity
import com.hustunique.morii.home.MusicDiaryAdapter.MyViewHolder
import com.hustunique.morii.util.DatabaseUtil
import com.hustunique.morii.util.MyApplication.Companion.externalPath
import com.hustunique.morii.util.MyApplication.Companion.musicTabList
import morii.R
import morii.databinding.CardviewItemBinding
import java.io.File


class MusicDiaryAdapter
    (private val activity: Activity, private val list: MutableList<MusicDiaryItem>) :
    RecyclerView.Adapter<MyViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val binding: CardviewItemBinding =
            CardviewItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return MyViewHolder(binding)
    }

    override fun onBindViewHolder(
        holder: MyViewHolder,
        @SuppressLint("RecyclerView") position: Int
    ) {
        val musicDiaryItem = list[position]
        musicDiaryItem.imagePath

        if (musicDiaryItem.imagePath != null)
            Glide.with(holder.itemView).load(musicDiaryItem.imagePath)
                .into(holder.itemBinding.PhotoTitle)
        else
            Glide.with(holder.itemView).load(musicTabList[musicDiaryItem.musicTabId].imageResId)
                .into(holder.itemBinding.PhotoTitle)

        holder.itemBinding.TextTitle.text = musicDiaryItem.title
        holder.itemBinding.TextDate.text = "# " + musicDiaryItem.date
        Log.d("RECYCLER", position.toString())
        holder.itemView.setOnClickListener {
            val intent = Intent(activity, ContentActivity::class.java)
            intent.putExtra("diary", musicDiaryItem)
            val options = ActivityOptions
                .makeSceneTransitionAnimation(
                    activity,
                    Pair.create(holder.itemBinding.PhotoTitle, "photo")
                )
            activity.startActivity(intent, options.toBundle())
        }
        holder.itemView.setOnLongClickListener {
            val builder = MaterialAlertDialogBuilder(activity)
            builder.run {
                setItems(arrayOf("操作 1", "操作 2", "删除日记")) { _, which ->
                    if (which != 2) {
                        Toast.makeText(activity, "Not yet implemented.", Toast.LENGTH_SHORT).show()
                    } else if (which == 2) {
                        MaterialAlertDialogBuilder(activity).setTitle("确定要删除这个音乐日记吗？")
                            .setMessage("这个操作不可被撤销。")
                            .setNegativeButton("取消") { _: DialogInterface?, _: Int -> }
                            .setPositiveButton("确认") { _: DialogInterface?, _: Int ->
                                DatabaseUtil.deleteDiary(musicDiaryItem.itemID)
                                DatabaseUtil.deleteAudioFile(
                                    musicDiaryItem.title,
                                    musicDiaryItem.date
                                )
                                Log.d(TAG, "deletePosition: " + holder.layoutPosition)
                                notifyItemRemoved(holder.layoutPosition)
                                list.removeAt(holder.layoutPosition)
                            }
                            .show()
                    }
                }
                show()
            }

            true
        }

    }

    override fun getItemCount(): Int {
        return list.size
    }

    //内部类，绑定控件
    inner class MyViewHolder(val itemBinding: CardviewItemBinding) :
        RecyclerView.ViewHolder(itemBinding.root)

    companion object {
        private const val TAG = "MusicDiaryAdapter"
    }
}